### PR TITLE
Adds a ParsingOptions OptionSetType

### LINKED ISSUE #128: Treat missing OR Null as nil
What if an incoming value in some JSON might be missing or might be Null. How do I allow both of those to be treated as nil. There are ifNotFound: and ifNull: variants of the subscripting functions, but I'm not sure how to allow both.

### PR DESCRIPTION
**Pull request by mdmathias**
`ParsingOptions` is used when parsing `JSON` that may have null values or missing keys.  `ParsingOptions provides` `.NullBecomesNil` and `.MissingKeyBecomesNil` options to treat this `JSON` as nil to prevent the throwing of errors.

This PR also adds tests for what Freddy considers `JSONSubscripting`.

Fixes #128

### CODE REVIEW

File: Sources/JSONSubscripting.swift [hunk 1]
```diff
@@ -214,50 +214,72 @@ extension JSON {
...
+    /// An `OptionSetType` used to represent the different options available for parsing `JSON` with `Null` values or missing keys.
+    /// * `.NullBecomesNil` - Treat `null` values as `nil`.
+    /// * `.MissingKeyBecomesNil` - Treat missing keys as `nil`.
+    public struct ParsingOptions: OptionSetType {
+        public let rawValue: Swift.Int
+        public init(rawValue: Swift.Int) {
+            self.rawValue = rawValue
+        }
+        
+        public static let NullBecomesNil = ParsingOptions(rawValue: 1)
```
Comment (jgallagher) [inline Sources/JSONSubscripting.swift]: Couple nitpicky things on these:

* They should have doc comments too. (Probably a copy-paste of what you have above is fine)
* Would prefer to see the raw values written as `1 << 0` and `1 << 1` so it's obvious if someone adds a third they need to say `1 << 2` and not `3`.

File: Sources/JSONSubscripting.swift [hunk 2]
```diff
@@ -214,50 +214,72 @@ extension JSON {
...
+    }
+    
+    private func mapOptionalAtPath<Value>(path: [JSONPathType], alongPath: ParsingOptions, @noescape transform: JSON throws -> Value) throws -> Value? {
+        let detectNull = alongPath.contains(.NullBecomesNil)
+        let detectNotFound = alongPath.contains(.MissingKeyBecomesNil)
+        var json: JSON?
         do {
-            return try valueAtPath(path, detectNull: true)
-        } catch Error.IndexOutOfBounds where ifNotFound {
+            json = try valueAtPath(path, detectNull: detectNull || detectNotFound)
```
Comment (jgallagher) [inline Sources/JSONSubscripting.swift]: `detectNull: detectNull || detectNotFound` doesn't seem right. If I take out `detectNotFound` here I get a failing test, but I'm not sure that test is right (will comment on it below).

File: Sources/JSONSubscripting.swift [hunk 3]
```diff
@@ -214,50 +214,72 @@ extension JSON {
...
+            return try json.map(transform)
+        } catch Error.IndexOutOfBounds where detectNotFound {
             return nil
-        } catch Error.KeyNotFound where ifNotFound {
+        } catch Error.KeyNotFound where detectNotFound {
             return nil
-        } catch SubscriptError.SubscriptIntoNull(_ as Swift.String) where ifNotFound {
+        } catch Error.ValueNotConvertible where detectNull && json == .Null {
+            return nil
+        } catch SubscriptError.SubscriptIntoNull where detectNull || detectNotFound {
```
Comment (jgallagher) [inline Sources/JSONSubscripting.swift]: Same as above - think `|| detectNotFound` doesn't seem right here.

File: Sources/JSONSubscripting.swift [hunk 4]
```diff
@@ -214,50 +214,72 @@ extension JSON {
...
 }
 
-// MARK: - Missing-to-Optional unpacking
+// MARK: - NotFound-Or-Null-to-Optional unpacking
 
 extension JSON {
-
-    private func optionalAtPath(path: [JSONPathType], ifNotFound: Swift.Bool) throws -> JSON? {
+    
+    /// An `OptionSetType` used to represent the different options available for parsing `JSON` with `Null` values or missing keys.
```
Comment (jgallagher) [inline Sources/JSONSubscripting.swift]: Super nitpick - `Null` should be `null`

File: Sources/JSONSubscripting.swift [hunk 5]
```diff
@@ -214,50 +214,72 @@ extension JSON {
...
+
+extension JSON {
 
     /// Optionally decodes into the returning type from a path into JSON.
     /// - parameter path: 0 or more `String` or `Int` that subscript the `JSON`
-    /// - parameter ifNotFound: If `true`, missing key or index errors are
-    ///                         treated as `nil`.
+    /// - parameter alongPath: An `OptionSetType` named `NullOptions` describing what should
+    ///                        be done with values that are `Null` or `NotFound`. Both
+    ///                        are treated as `nil`.
```
Comment (jgallagher) [inline Sources/JSONSubscripting.swift]: The docs for this don't need to be so specific about the type (that's already obvious from the signature). Maybe:

```
- parameter alongPath: Options controlling what should be done with values that are `null` or missing.
```

Not sure what "Both are treated as `nil`" means?

Comment (jgallagher) [inline Sources/JSONSubscripting.swift]: (not repeating this comment, but any changes should be copied to all the places where this comment is made)

File: Tests/JSONSubscriptingTests.swift [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+    }
+    
+    func testNullOptionsProducesOptionalForNotFound() {
+        do {
+            let residents = try json.arrayOf("residents", type: Resident.self)
+            XCTAssertNil(residents[1].age, "Drew's `age` should be nil.")
+        } catch {
+            XCTFail("There should be no error.")
+        }
+    }
```
Comment (jgallagher) [inline Tests/JSONSubscriptingTests.swift]: I'd add a couple more tests that when one of the nil-like options is found but the data has the other, an error is still thrown. Something like:

```swift
    func testMissingKeyOptionStillFailsIfNullEncountered() {
        do {
            let json = JSON.Dictionary([
                "name": "Drew",
                "age": nil, // should cause problems!
                "hasPet": true,
                "rent": 1234.5,
            ])
            let resident = try Resident(json: json)
            XCTFail("Unexpected success: \(resident)")
        } catch let JSON.Error.ValueNotConvertible(value: value, to: type) {
            XCTAssert(type == Int.self, "unexpected type \(type) of value \(value)")
        } catch {
            XCTFail("Unexpected error: \(error)")
        }
    }

    func testNullOptionsStillFailIfKeyIsMissing() {
        do {
            let json = JSON.Dictionary([
                "name": "Drew",
                "rent": 1234.5,
            ])
            let resident = try Resident(json: json)
            XCTFail("Unexpected success: \(resident)")
        } catch let JSON.Error.KeyNotFound(key: key) where key == "hasPet" {
            // expected
        } catch {
            XCTFail("Unexpected error: \(error)")
        }
    }
```

File: Tests/JSONSubscriptingTests.swift [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+            XCTFail("This should not be: \(error).")
+        }
+    }
+    
+    func testSubscriptIntoNullNotFound() {
+        do {
+            let nullJSON: JSON = .Null
+            _ = try nullJSON.string("path", alongPath: .MissingKeyBecomesNil)
+        } catch {
+            XCTFail("There should be no error: \(error).")
```
Comment (jgallagher) [inline Tests/JSONSubscriptingTests.swift]: This test doesn't seem right to me. I would've expected this to pass with `.NullBecomesNil`, but it seems like `.MissingKeyBecomesNil` implies we have a dictionary which is missing some keys.

File: Sources/JSONSubscripting.swift (hunk 4 above)
Comment (mdmathias) [inline Sources/JSONSubscripting.swift]: :+1:

File: Sources/JSONSubscripting.swift (hunk 1 above)
Comment (mdmathias) [inline Sources/JSONSubscripting.swift]: Yep.  Thanks!

File: Sources/JSONSubscripting.swift (hunk 5 above)
Comment (mdmathias) [inline Sources/JSONSubscripting.swift]: "Both are treated as `nil`" is intended to mean that this is what you get out if either is encountered.  I guess that isn't needed here because it is in the documentation for the return.

File: Tests/JSONSubscriptingTests.swift (hunk 6 above)
Comment (mdmathias) [inline Tests/JSONSubscriptingTests.swift]: :+1:

File: Tests/JSONSubscriptingTests.swift (hunk 7 above)
Comment (mdmathias) [inline Tests/JSONSubscriptingTests.swift]: I was being a little lazy with `nullJSON`. I'm trying to test the last `catch` here:

```swift
private func mapOptionalAtPath<Value>(path: [JSONPathType], alongPath: ParsingOptions, @noescape transform: JSON throws -> Value) throws -> Value? {
    let detectNull = alongPath.contains(.NullBecomesNil)
    let detectNotFound = alongPath.contains(.MissingKeyBecomesNil)
    var json: JSON?
    do {
        json = try valueAtPath(path, detectNull: detectNull || detectNotFound)
        return try json.map(transform)
    } catch Error.IndexOutOfBounds where detectNotFound {
        return nil
    } catch Error.KeyNotFound where detectNotFound {
        return nil
    } catch Error.ValueNotConvertible where detectNull && json == .Null {
        return nil
    } catch SubscriptError.SubscriptIntoNull where detectNull || detectNotFound {
        return nil
    }
}
```

Would you prefer something like this?

```swift
func testSubscriptIntoNullNotFound() {
    do {
        let sampleJSON: JSON = ["results": .Null]
        _ = try sampleJSON.dictionary("results", "users", alongPath: .MissingKeyBecomesNil)
    } catch {
        XCTFail("There should be no error: \(error).")
    }
}
```

File: Sources/JSONSubscripting.swift (hunk 2 above)
Comment (mdmathias) [inline Sources/JSONSubscripting.swift]: That test will fail because of an unexpected subscript on a `JSON.Null`.  Is it better to add `catch` for this scenario (see the last `catch`)?

```swift
private func mapOptionalAtPath<Value>(path: [JSONPathType], alongPath: ParsingOptions, @noescape transform: JSON throws -> Value) throws -> Value? {
    let detectNull = alongPath.contains(.NullBecomesNil)
    let detectNotFound = alongPath.contains(.MissingKeyBecomesNil)
    var json: JSON?
    do {
        json = try valueAtPath(path, detectNull: detectNull)
        return try json.map(transform)
    } catch Error.IndexOutOfBounds where detectNotFound {
        return nil
    } catch Error.KeyNotFound where detectNotFound {
        return nil
    } catch Error.ValueNotConvertible where detectNull && json == .Null {
        return nil
    } catch SubscriptError.SubscriptIntoNull where detectNull || detectNotFound {
        return nil
    } 
    catch Error.UnexpectedSubscript where detectNotFound {
        return nil
    }
}
```

Or, perhaps you're saying that there should be no `catch` like the above and this _should_ be an error if code like the test referenced below is written?

Comment (mdmathias) [inline Sources/JSONSubscripting.swift]: Upon further reflection, that last `catch` I added above feels silly.

File: Sources/JSONSubscripting.swift (hunk 3 above)
Comment (mdmathias) [inline Sources/JSONSubscripting.swift]: I agree with you.  What does `detectNotFound` mean when you have subscripted into `.Null`?  I do not know.

This is what I have now:

```swift
private func mapOptionalAtPath<Value>(path: [JSONPathType], alongPath: ParsingOptions, @noescape transform: JSON throws -> Value) throws -> Value? {
    let detectNull = alongPath.contains(.NullBecomesNil)
    let detectNotFound = alongPath.contains(.MissingKeyBecomesNil)
    var json: JSON?
    do {
        json = try valueAtPath(path, detectNull: detectNull)
        return try json.map(transform)
    } catch Error.IndexOutOfBounds where detectNotFound {
        return nil
    } catch Error.KeyNotFound where detectNotFound {
        return nil
    } catch Error.ValueNotConvertible where detectNull && json == .Null {
        return nil
    } catch SubscriptError.SubscriptIntoNull where detectNull {
        return nil
    }
}
```

And that failing test is now passing and looks like:

```swift
func testSubscriptIntoNullBecomesNil() {
        do {
            let sampleJSON: JSON = ["results": .Null]
            _ = try sampleJSON.dictionary("results", "users", alongPath: .NullBecomesNil)
        } catch {
            XCTFail("There should be no error: \(error).")
        }
    }
```

But it doesn't increase our test coverage; so, I think it's not necessary.

File: Tests/JSONSubscriptingTests.swift (hunk 7 above)
Comment (jgallagher) [inline Tests/JSONSubscriptingTests.swift]: Yes, I like that test much better. :+1:

One other note about the tests - the last catch is tripped by `testThatOptionalSubscriptingIntoNullSucceeds()` in `JSONSubscriptTests.swift`, which brings up another question - we now have `JSONSubscriptTests.swift` and `JSONSubscriptingTests.swift`. Are those testing different things, or should they be combined somehow?

File: Sources/JSONSubscripting.swift [hunk 8]
```diff
@@ -214,50 +214,70 @@ extension JSON {
...
+        } catch Error.IndexOutOfBounds where detectNotFound {
             return nil
-        } catch Error.KeyNotFound where ifNotFound {
+        } catch Error.KeyNotFound where detectNotFound {
             return nil
-        } catch SubscriptError.SubscriptIntoNull(_ as Swift.String) where ifNotFound {
+        } catch Error.ValueNotConvertible where detectNull && json == .Null {
+            return nil
+        } catch SubscriptError.SubscriptIntoNull where detectNull {
             return nil
```
Comment (zwaldowski) [inline Sources/JSONSubscripting.swift]: There should be another `catch SubscriptError.SubscriptIntoNull` that rethrows as one of our other errors, I think

File: Sources/JSONSubscripting.swift [hunk 9]
```diff
@@ -214,50 +214,70 @@ extension JSON {
...
+// MARK: - NotFound-Or-Null-to-Optional unpacking
 
 extension JSON {
-
-    private func optionalAtPath(path: [JSONPathType], ifNotFound: Swift.Bool) throws -> JSON? {
+    
+    /// An `OptionSetType` used to represent the different options available for parsing `JSON` with `null` values or missing keys.
+    /// * `.NullBecomesNil` - Treat `null` values as `nil`.
+    /// * `.MissingKeyBecomesNil` - Treat missing keys as `nil`.
+    public struct ParsingOptions: OptionSetType {
```
Comment (zwaldowski) [inline Sources/JSONSubscripting.swift]: Aren't these more `TraversalOptions` (or sth.) than `ParsingOptions`?

File: Tests/JSONSubscriptingTests.swift (hunk 7 above)
Comment (mdmathias) [inline Tests/JSONSubscriptingTests.swift]: :+1: 

I don't think they are testing different things.  I'll combine them.

File: Sources/JSONSubscripting.swift (hunk 8 above)
Comment (mdmathias) [inline Sources/JSONSubscripting.swift]: I don't think that `catch` will ever be hit given:

```swift
func valueForPathFragment(fragment: JSONPathType, detectNull: Swift.Bool) throws -> JSON {
    switch self {
    case .Null where detectNull:
        throw SubscriptError.SubscriptIntoNull(fragment)
    case let .Dictionary(dict):
        return try fragment.valueInDictionary(dict)
    case let .Array(array):
        return try fragment.valueInArray(array)
    default:
        throw Error.UnexpectedSubscript(type: fragment.dynamicType)
    }
}
```

In the current implementation, the `default` above will match, and `Error.UnexpectedSubscript` will be thrown if somebody uses this `mapOptionalAtPath` without specifying `.NullBecomesNil`.

File: Sources/JSONSubscripting.swift (hunk 9 above)
Comment (mdmathias) [inline Sources/JSONSubscripting.swift]: I don't think of this as dictating how the `JSON` is traversed.  What about `SubscriptingOptions`?

File: Sources/JSONSubscripting.swift (hunk 8 above)
Comment (zwaldowski) [inline Sources/JSONSubscripting.swift]: I'll test to make sure but :+1:. Just don't want to publish a private error

File: Sources/JSONSubscripting.swift (hunk 9 above)
Comment (zwaldowski) [inline Sources/JSONSubscripting.swift]: Isn't it? The subscript is a traversal.

(I don't really care, whatever you think is fine, I just don't think `Parsing` describes what's going on. :smiley:)

Comment (mdmathias) [inline Sources/JSONSubscripting.swift]: Maybe you're right.  The way I am thinking about it is that traversal describes the direction a path or route takes.  I don't see these options as describing how a path should be formed; rather, I see these options as describing what to do if a particular thing is found (`null`) or not found (a key).  Hence, "if you find a `null`, parse that into a `nil` for me.  Or, "if you don't find this key, parse that into `nil` as well.

Does that sound wrong or misleading?

In any case, I kind of like `SubscriptingOptions` better.  I'll make that change and you let me know if you think that sounds better or worse.

Comment (zwaldowski) [inline Sources/JSONSubscripting.swift]: Nope, you're 100% right. Change away.

**Comment by zwaldowski** [comment]
We should get this in soon. @jgallagher pointed out the other day that this, a breaking API change, means a x.0.0 bump based on strict adherence to semver.

* Do we consider adding backwards compatible method definitions, which would instead be a 0.x.0 bump?
* As Swift resilience isn't, well, a thing yet, do we not strictly to adhere to semver this early on in Freddy's life?

**Comment by jgallagher** [comment]
I would rather adhere to semver correctly. Adding deprecated backwards compatible methods sounds best to me.

**Comment by zwaldowski** [comment]
Great! Swift gives us both forwarding/`@available(*, deprecated)` versions and `@available(*, unavailable)` versions. Either choice we make, I feel that, as long as we are careful with the copypasta, the backwards versions don't need the full docs and testing treatment.

I'll add this to the Milestone; "Next" will tentatively be 2.1.

**Comment by mdmathias** [comment]
So, shall we merge?

**Comment by zwaldowski** [comment]
We need to add the `ifNull`-parameter versions back for compat, but that can be done separately. I'll file an issue in my morning. :sleeping: otherwise we're good

**Comment by mdmathias** [comment]
:zzz: well!
